src/Futhark/LSP/PositionMapping.hs

```diff
@@ -0,0 +1,72 @@
+-- | Provide mapping between position in stale content and current.
+module Futhark.LSP.PositionMapping
+  ( mappingFromDiff,
+    PositionMapping,
+    toStalePos,
+    toCurrentLoc,
+    StaleFile (..),
+  )
+where
+
+import Data.Algorithm.Diff (Diff, PolyDiff (Both, First, Second), getDiff)
+import Data.Bifunctor (Bifunctor (bimap, first, second))
+import qualified Data.Text as T
+import Futhark.Util.Loc (Loc (Loc), Pos (Pos))
+import Language.LSP.VFS (VirtualFile)
+
+-- | A mapping between current file content and the stale (last successful compiled) file content
+-- currently, only supports entire line mapping
+-- more detailed mapping might be achieved via referring to haskell-language-server@efb4b94
+newtype PositionMapping = PositionMapping ([Int], [Int])
```

Comment:
I am not comfortable using lists here, as lookups are slow *especially* if at some point we start doing many lookups (e.g. for semantic highlighting).  Use a Vector instead, like I think you have done elsewhere.  Also, document what the two sequences actually mean.

Reply:
Good point